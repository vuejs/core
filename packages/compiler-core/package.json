{
  "name": "@vue/compiler-core",
  "version": "3.5.13",
  "description": "@vue/compiler-core",
  "main": "index.js",
  "module": "dist/compiler-core.esm-bundler.js",
  "types": "dist/compiler-core.d.ts",
  "files": [
    "index.js",
    "dist"
  ],
  "exports": {
    ".": {
      "types": "./dist/compiler-core.d.ts",
      "node": {
        "production": "./dist/compiler-core.cjs.prod.js",
        "development": "./dist/compiler-core.cjs.js",
        "default": "./index.js"
      },
      "module": "./dist/compiler-core.esm-bundler.js",
      "import": "./dist/compiler-core.esm-bundler.js",
      "require": "./index.js"
    },
    "./*": "./*"
  },
  "buildOptions": {
    "name": "VueCompilerCore",
    "compat": true,
    "formats": [
      "esm-bundler",
      "cjs"
    ]
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vuejs/core.git",
    "directory": "packages/compiler-core"
  },
  "keywords": [
    "vue"
  ],
  "author": "Evan You",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/vuejs/core/issues"
  },
  "homepage": "https://github.com/vuejs/core/tree/main/packages/compiler-core#readme",
  "dependencies": {
    "@babel/parser": "catalog:",
    "@vue/shared": "workspace:*",
    "entities": "^4.5.0",
    "estree-walker": "catalog:",
    "source-map-js": "catalog:"
  },
  "devDependencies": {
    "@babel/types": "catalog:"
  }
}
