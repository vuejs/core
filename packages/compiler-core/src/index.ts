export { baseCompile } from './compile'

// Also expose lower level APIs & types
export {
  type CompilerOptions,
  type ParserOptions,
  type TransformOptions,
  type CodegenOptions,
  type HoistTransform,
  type BindingMetadata,
  BindingTypes,
} from './options'
export { baseParse } from './parser'
export {
  transform,
  type TransformContext,
  createTransformContext,
  traverseNode,
  createStructuralDirectiveTransform,
  type NodeTransform,
  type StructuralDirectiveTransform,
  type DirectiveTransform,
} from './transform'
export {
  generate,
  type CodegenContext,
  type CodegenResult,
  type CodegenSourceMapGenerator,
  type RawSourceMap,
} from './codegen'
export {
  ErrorCodes,
  errorMessages,
  createCompilerError,
  type CoreCompilerError,
  type CompilerError,
} from './errors'

export * from './ast'
export * from './utils'
export * from './babelUtils'
export * from './runtimeHelpers'

export { getBaseTransformPreset, type TransformPreset } from './compile'
export { transformModel } from './transforms/vModel'
export { transformOn } from './transforms/vOn'
export { transformBind } from './transforms/vBind'
export { noopDirectiveTransform } from './transforms/noopDirectiveTransform'
export { processIf } from './transforms/vIf'
export { processFor, createForLoopParams } from './transforms/vFor'
export {
  transformExpression,
  processExpression,
  stringifyExpression,
} from './transforms/transformExpression'
export {
  buildSlots,
  type SlotFnBuilder,
  trackVForSlotScopes,
  trackSlotScopes,
} from './transforms/vSlot'
export {
  transformElement,
  resolveComponentType,
  buildProps,
  buildDirectiveArgs,
  type PropsExpression,
} from './transforms/transformElement'
export { processSlotOutlet } from './transforms/transformSlotOutlet'
export { getConstantType } from './transforms/cacheStatic'
export { generateCodeFrame } from '@vue/shared'

// v2 compat only
export {
  checkCompatEnabled,
  warnDeprecation,
  CompilerDeprecationTypes,
} from './compat/compatConfig'
