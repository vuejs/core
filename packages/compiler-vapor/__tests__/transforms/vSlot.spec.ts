import { ErrorCodes, NodeTypes } from '@vue/compiler-core'
import {
  IRNodeTypes,
  IRSlotType,
  transformChildren,
  transformElement,
  transformSlotOutlet,
  transformText,
  transformVBind,
  transformVFor,
  transformVIf,
  transformVOn,
  transformVSlot,
} from '../../src'
import { makeCompile } from './_utils'

const compileWithSlots = makeCompile({
  nodeTransforms: [
    transformText,
    transformVIf,
    transformVFor,
    transformSlotOutlet,
    transformElement,
    transformVSlot,
    transformChildren,
  ],
  directiveTransforms: {
    bind: transformVBind,
    on: transformVOn,
  },
})

describe('compiler: transform slot', () => {
  test('implicit default slot', () => {
    const { ir, code } = compileWithSlots(`<Comp><div/></Comp>`)
    expect(code).toMatchSnapshot()

    expect(ir.template).toEqual(['<div></div>'])
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      id: 1,
      tag: 'Comp',
      props: [[]],
      slots: [
        {
          slotType: IRSlotType.STATIC,
          slots: {
            default: {
              type: IRNodeTypes.BLOCK,
              dynamic: {
                children: [{ template: 0 }],
              },
            },
          },
        },
      ],
    })
    expect(ir.block.returns).toEqual([1])
    expect(ir.block.dynamic).toMatchObject({
      children: [{ id: 1 }],
    })
  })

  test('on-component default slot', () => {
    const { ir, code } = compileWithSlots(
      `<Comp v-slot="{ foo }">{{ foo + bar }}</Comp>`,
    )
    expect(code).toMatchSnapshot()

    expect(code).contains(`"default": (_slotProps0) =>`)
    expect(code).contains(`_slotProps0["foo"] + _ctx.bar`)

    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      tag: 'Comp',
      props: [[]],
      slots: [
        {
          slotType: IRSlotType.STATIC,
          slots: {
            default: {
              type: IRNodeTypes.BLOCK,
              props: {
                type: NodeTypes.SIMPLE_EXPRESSION,
                content: '{ foo }',
                ast: {
                  type: 'ArrowFunctionExpression',
                  params: [{ type: 'ObjectPattern' }],
                },
              },
            },
          },
        },
      ],
    })
  })

  test('on component named slot', () => {
    const { ir, code } = compileWithSlots(
      `<Comp v-slot:named="{ foo }">{{ foo + bar }}</Comp>`,
    )
    expect(code).toMatchSnapshot()

    expect(code).contains(`"named": (_slotProps0) =>`)
    expect(code).contains(`_slotProps0["foo"] + _ctx.bar`)

    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      tag: 'Comp',
      slots: [
        {
          slotType: IRSlotType.STATIC,
          slots: {
            named: {
              type: IRNodeTypes.BLOCK,
              props: {
                type: NodeTypes.SIMPLE_EXPRESSION,
                content: '{ foo }',
              },
            },
          },
        },
      ],
    })
  })

  test('on component dynamically named slot', () => {
    const { ir, code } = compileWithSlots(
      `<Comp v-slot:[named]="{ foo }">{{ foo + bar }}</Comp>`,
    )
    expect(code).toMatchSnapshot()

    expect(code).contains(`fn: (_slotProps0) =>`)
    expect(code).contains(`_slotProps0["foo"] + _ctx.bar`)

    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      tag: 'Comp',
      slots: [
        {
          name: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'named',
            isStatic: false,
          },
          fn: {
            type: IRNodeTypes.BLOCK,
            props: {
              type: NodeTypes.SIMPLE_EXPRESSION,
              content: '{ foo }',
            },
          },
        },
      ],
    })
  })

  test('named slots w/ implicit default slot', () => {
    const { ir, code } = compileWithSlots(
      `<Comp>
        <template #one>foo</template>bar<span/>
      </Comp>`,
    )
    expect(code).toMatchSnapshot()

    expect(ir.template).toEqual(['foo', 'bar', '<span></span>'])
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      id: 4,
      tag: 'Comp',
      props: [[]],
      slots: [
        {
          slotType: IRSlotType.STATIC,
          slots: {
            one: {
              type: IRNodeTypes.BLOCK,
              dynamic: {
                children: [{ template: 0 }],
              },
            },
            default: {
              type: IRNodeTypes.BLOCK,
              dynamic: {
                children: [{}, { template: 1 }, { template: 2 }],
              },
            },
          },
        },
      ],
    })
  })

  test('nested slots scoping', () => {
    const { ir, code } = compileWithSlots(
      `<Comp>
        <template #default="{ foo }">
          <Inner v-slot="{ bar }">
            {{ foo + bar + baz }}
          </Inner>
          {{ foo + bar + baz }}
        </template>
      </Comp>`,
    )
    expect(code).toMatchSnapshot()

    expect(code).contains(`"default": (_slotProps0) =>`)
    expect(code).contains(`"default": (_slotProps1) =>`)
    expect(code).contains(`_slotProps0["foo"] + _slotProps1["bar"] + _ctx.baz`)
    expect(code).contains(`_slotProps0["foo"] + _ctx.bar + _ctx.baz`)

    const outerOp = ir.block.dynamic.children[0].operation
    expect(outerOp).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      tag: 'Comp',
      props: [[]],
      slots: [
        {
          slotType: IRSlotType.STATIC,
          slots: {
            default: {
              type: IRNodeTypes.BLOCK,
              props: {
                type: NodeTypes.SIMPLE_EXPRESSION,
                content: '{ foo }',
              },
            },
          },
        },
      ],
    })
    expect(
      (outerOp as any).slots[0].slots.default.dynamic.children[0].operation,
    ).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      tag: 'Inner',
      slots: [
        {
          slotType: IRSlotType.STATIC,
          slots: {
            default: {
              type: IRNodeTypes.BLOCK,
              props: {
                type: NodeTypes.SIMPLE_EXPRESSION,
                content: '{ bar }',
              },
            },
          },
        },
      ],
    })
  })

  test('dynamic slots name', () => {
    const { ir, code } = compileWithSlots(
      `<Comp>
        <template #[name]>foo</template>
      </Comp>`,
    )
    expect(code).toMatchSnapshot()
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      tag: 'Comp',
      slots: [
        {
          name: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'name',
            isStatic: false,
          },
          fn: { type: IRNodeTypes.BLOCK },
        },
      ],
    })
  })

  test('dynamic slots name w/ v-for', () => {
    const { ir, code } = compileWithSlots(
      `<Comp>
        <template v-for="item in list" #[item]="{ bar }">{{ bar }}</template>
      </Comp>`,
    )
    expect(code).toMatchSnapshot()

    expect(code).contains(`fn: (_slotProps0) =>`)
    expect(code).contains(`_setText(n0, _toDisplayString(_slotProps0["bar"]))`)

    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      tag: 'Comp',
      slots: [
        {
          name: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'item',
            isStatic: false,
          },
          fn: { type: IRNodeTypes.BLOCK },
          loop: {
            source: { content: 'list' },
            value: { content: 'item' },
            index: undefined,
          },
        },
      ],
    })
  })

  test('dynamic slots name w/ v-for and provide absent key', () => {
    const { ir, code } = compileWithSlots(
      `<Comp>
        <template v-for="(,,index) in list" #[index]>foo</template>
      </Comp>`,
    )
    expect(code).toMatchSnapshot()
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      tag: 'Comp',
      slots: [
        {
          name: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'index',
            isStatic: false,
          },
          fn: { type: IRNodeTypes.BLOCK },
          loop: {
            source: { content: 'list' },
            value: undefined,
            index: {
              type: NodeTypes.SIMPLE_EXPRESSION,
            },
          },
        },
      ],
    })
  })

  test('dynamic slots name w/ v-if / v-else[-if]', () => {
    const { ir, code } = compileWithSlots(
      `<Comp>
        <template v-if="condition" #condition>condition slot</template>
        <template v-else-if="anotherCondition" #condition="{ foo, bar }">another condition</template>
        <template v-else #condition>else condition</template>
      </Comp>`,
    )
    expect(code).toMatchSnapshot()

    expect(code).contains(`fn: (_slotProps0) =>`)

    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      tag: 'Comp',
      slots: [
        {
          slotType: IRSlotType.CONDITIONAL,
          condition: { content: 'condition' },
          positive: {
            slotType: IRSlotType.DYNAMIC,
          },
          negative: {
            slotType: IRSlotType.CONDITIONAL,
            condition: { content: 'anotherCondition' },
            positive: {
              slotType: IRSlotType.DYNAMIC,
            },
            negative: { slotType: IRSlotType.DYNAMIC },
          },
        },
      ],
    })
  })

  test('slot + v-if / v-else[-if] should not cause error', () => {
    const { code } = compileWithSlots(
      `<div>
        <slot name="foo"></slot>
        <Foo v-if="true"></Foo>
        <Bar v-else />
      </div>`,
    )
    expect(code).toMatchSnapshot()
  })

  test('quote slot name', () => {
    const { code } = compileWithSlots(
      `<Comp><template #nav-bar-title-before></template></Comp>`,
    )
    expect(code).toMatchSnapshot()
    expect(code).contains(`"nav-bar-title-before"`)
  })

  test('nested component slot', () => {
    const { ir, code } = compileWithSlots(`<A><B/></A>`)
    expect(code).toMatchSnapshot()
    expect(ir.block.dynamic.children[0].operation).toMatchObject({
      type: IRNodeTypes.CREATE_COMPONENT_NODE,
      tag: 'A',
      slots: [
        {
          slotType: IRSlotType.STATIC,
          slots: {
            default: {
              type: IRNodeTypes.BLOCK,
              dynamic: {
                children: [
                  {
                    operation: {
                      type: IRNodeTypes.CREATE_COMPONENT_NODE,
                      tag: 'B',
                      slots: [],
                    },
                  },
                ],
              },
            },
          },
        },
      ],
    })
  })

  describe('errors', () => {
    test('error on extraneous children w/ named default slot', () => {
      const onError = vi.fn()
      const source = `<Comp><template #default>foo</template>bar</Comp>`
      compileWithSlots(source, { onError })
      const index = source.indexOf('bar')
      expect(onError.mock.calls[0][0]).toMatchObject({
        code: ErrorCodes.X_V_SLOT_EXTRANEOUS_DEFAULT_SLOT_CHILDREN,
        loc: {
          start: {
            offset: index,
            line: 1,
            column: index + 1,
          },
          end: {
            offset: index + 3,
            line: 1,
            column: index + 4,
          },
        },
      })
    })

    test('error on duplicated slot names', () => {
      const onError = vi.fn()
      const source = `<Comp><template #foo></template><template #foo></template></Comp>`
      compileWithSlots(source, { onError })
      const index = source.lastIndexOf('#foo')
      expect(onError.mock.calls[0][0]).toMatchObject({
        code: ErrorCodes.X_V_SLOT_DUPLICATE_SLOT_NAMES,
        loc: {
          start: {
            offset: index,
            line: 1,
            column: index + 1,
          },
          end: {
            offset: index + 4,
            line: 1,
            column: index + 5,
          },
        },
      })
    })

    test('error on invalid mixed slot usage', () => {
      const onError = vi.fn()
      const source = `<Comp v-slot="foo"><template #foo></template></Comp>`
      compileWithSlots(source, { onError })
      const index = source.lastIndexOf('v-slot="foo"')
      expect(onError.mock.calls[0][0]).toMatchObject({
        code: ErrorCodes.X_V_SLOT_MIXED_SLOT_USAGE,
        loc: {
          start: {
            offset: index,
            line: 1,
            column: index + 1,
          },
          end: {
            offset: index + 12,
            line: 1,
            column: index + 13,
          },
        },
      })
    })

    test('error on v-slot usage on plain elements', () => {
      const onError = vi.fn()
      const source = `<div v-slot/>`
      compileWithSlots(source, { onError })
      const index = source.indexOf('v-slot')
      expect(onError.mock.calls[0][0]).toMatchObject({
        code: ErrorCodes.X_V_SLOT_MISPLACED,
        loc: {
          start: {
            offset: index,
            line: 1,
            column: index + 1,
          },
          end: {
            offset: index + 6,
            line: 1,
            column: index + 7,
          },
        },
      })
    })
  })
})
