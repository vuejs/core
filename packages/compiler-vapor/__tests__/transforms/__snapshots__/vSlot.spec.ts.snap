// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: transform slot > dynamic slots name 1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback, template as _template } from 'vue';
const t0 = _template("foo")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n2 = _createComponentWithFallback(_component_Comp, null, {
    $: [
      () => ({
        name: _ctx.name, 
        fn: () => {
          const n0 = t0()
          return n0
        }
      })
    ]
  }, true)
  return n2
}"
`;

exports[`compiler: transform slot > dynamic slots name w/ v-for 1`] = `
"import { resolveComponent as _resolveComponent, toDisplayString as _toDisplayString, setText as _setText, renderEffect as _renderEffect, createForSlots as _createForSlots, createComponentWithFallback as _createComponentWithFallback, template as _template } from 'vue';
const t0 = _template(" ")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n2 = _createComponentWithFallback(_component_Comp, null, {
    $: [
      () => (_createForSlots(_ctx.list, (item) => ({
        name: item, 
        fn: (_slotProps0) => {
          const n0 = t0()
          _renderEffect(() => _setText(n0, _toDisplayString(_slotProps0["bar"])))
          return n0
        }
      })))
    ]
  }, true)
  return n2
}"
`;

exports[`compiler: transform slot > dynamic slots name w/ v-for and provide absent key 1`] = `
"import { resolveComponent as _resolveComponent, createForSlots as _createForSlots, createComponentWithFallback as _createComponentWithFallback, template as _template } from 'vue';
const t0 = _template("foo")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n2 = _createComponentWithFallback(_component_Comp, null, {
    $: [
      () => (_createForSlots(_ctx.list, (_, __, index) => ({
        name: index, 
        fn: () => {
          const n0 = t0()
          return n0
        }
      })))
    ]
  }, true)
  return n2
}"
`;

exports[`compiler: transform slot > dynamic slots name w/ v-if / v-else[-if] 1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback, template as _template } from 'vue';
const t0 = _template("condition slot")
const t1 = _template("another condition")
const t2 = _template("else condition")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n6 = _createComponentWithFallback(_component_Comp, null, {
    $: [
      () => (_ctx.condition
        ? {
          name: "condition", 
          fn: () => {
            const n0 = t0()
            return n0
          }
        }
        : _ctx.anotherCondition
          ? {
            name: "condition", 
            fn: (_slotProps0) => {
              const n2 = t1()
              return n2
            }
          }
          : {
            name: "condition", 
            fn: () => {
              const n4 = t2()
              return n4
            }
          })
    ]
  }, true)
  return n6
}"
`;

exports[`compiler: transform slot > implicit default slot 1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback, template as _template } from 'vue';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n1 = _createComponentWithFallback(_component_Comp, null, {
    "default": () => {
      const n0 = t0()
      return n0
    }
  }, true)
  return n1
}"
`;

exports[`compiler: transform slot > named slots w/ implicit default slot 1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback, template as _template } from 'vue';
const t0 = _template("foo")
const t1 = _template("bar")
const t2 = _template("<span></span>")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n4 = _createComponentWithFallback(_component_Comp, null, {
    "one": () => {
      const n0 = t0()
      return n0
    }, 
    "default": () => {
      const n2 = t1()
      const n3 = t2()
      return [n2, n3]
    }
  }, true)
  return n4
}"
`;

exports[`compiler: transform slot > nested component slot 1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback } from 'vue';

export function render(_ctx) {
  const _component_B = _resolveComponent("B")
  const _component_A = _resolveComponent("A")
  const n1 = _createComponentWithFallback(_component_A, null, {
    "default": () => {
      const n0 = _createComponentWithFallback(_component_B)
      return n0
    }
  }, true)
  return n1
}"
`;

exports[`compiler: transform slot > nested slots scoping 1`] = `
"import { resolveComponent as _resolveComponent, toDisplayString as _toDisplayString, setText as _setText, renderEffect as _renderEffect, createComponentWithFallback as _createComponentWithFallback, template as _template } from 'vue';
const t0 = _template(" ")

export function render(_ctx) {
  const _component_Inner = _resolveComponent("Inner")
  const _component_Comp = _resolveComponent("Comp")
  const n5 = _createComponentWithFallback(_component_Comp, null, {
    "default": (_slotProps0) => {
      const n1 = _createComponentWithFallback(_component_Inner, null, {
        "default": (_slotProps1) => {
          const n0 = t0()
          _renderEffect(() => _setText(n0, _toDisplayString(_slotProps0["foo"] + _slotProps1["bar"] + _ctx.baz)))
          return n0
        }
      })
      const n3 = t0()
      _renderEffect(() => _setText(n3, " " + _toDisplayString(_slotProps0["foo"] + _ctx.bar + _ctx.baz)))
      return [n1, n3]
    }
  }, true)
  return n5
}"
`;

exports[`compiler: transform slot > on component dynamically named slot 1`] = `
"import { resolveComponent as _resolveComponent, toDisplayString as _toDisplayString, setText as _setText, renderEffect as _renderEffect, createComponentWithFallback as _createComponentWithFallback, template as _template } from 'vue';
const t0 = _template(" ")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n1 = _createComponentWithFallback(_component_Comp, null, {
    $: [
      () => ({
        name: _ctx.named, 
        fn: (_slotProps0) => {
          const n0 = t0()
          _renderEffect(() => _setText(n0, _toDisplayString(_slotProps0["foo"] + _ctx.bar)))
          return n0
        }
      })
    ]
  }, true)
  return n1
}"
`;

exports[`compiler: transform slot > on component named slot 1`] = `
"import { resolveComponent as _resolveComponent, toDisplayString as _toDisplayString, setText as _setText, renderEffect as _renderEffect, createComponentWithFallback as _createComponentWithFallback, template as _template } from 'vue';
const t0 = _template(" ")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n1 = _createComponentWithFallback(_component_Comp, null, {
    "named": (_slotProps0) => {
      const n0 = t0()
      _renderEffect(() => _setText(n0, _toDisplayString(_slotProps0["foo"] + _ctx.bar)))
      return n0
    }
  }, true)
  return n1
}"
`;

exports[`compiler: transform slot > on-component default slot 1`] = `
"import { resolveComponent as _resolveComponent, toDisplayString as _toDisplayString, setText as _setText, renderEffect as _renderEffect, createComponentWithFallback as _createComponentWithFallback, template as _template } from 'vue';
const t0 = _template(" ")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n1 = _createComponentWithFallback(_component_Comp, null, {
    "default": (_slotProps0) => {
      const n0 = t0()
      _renderEffect(() => _setText(n0, _toDisplayString(_slotProps0["foo"] + _ctx.bar)))
      return n0
    }
  }, true)
  return n1
}"
`;

exports[`compiler: transform slot > quote slot name 1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback } from 'vue';

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n1 = _createComponentWithFallback(_component_Comp, null, {
    "nav-bar-title-before": () => {
      return null
    }
  }, true)
  return n1
}"
`;

exports[`compiler: transform slot > slot + v-if / v-else[-if] should not cause error 1`] = `
"import { resolveComponent as _resolveComponent, setInsertionState as _setInsertionState, createSlot as _createSlot, createComponentWithFallback as _createComponentWithFallback, createIf as _createIf, template as _template } from 'vue';
const t0 = _template("<div></div>", true)

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const _component_Bar = _resolveComponent("Bar")
  const n6 = t0()
  _setInsertionState(n6)
  const n0 = _createSlot("foo", null)
  _setInsertionState(n6)
  const n1 = _createIf(() => (true), () => {
    const n3 = _createComponentWithFallback(_component_Foo)
    return n3
  }, () => {
    const n5 = _createComponentWithFallback(_component_Bar)
    return n5
  })
  return n6
}"
`;
